### PR TITLE
[MLIR] extend `getCompressedMaskOp` support in `VectorEmulateNarrowType`

### PR DESCRIPTION
**Pull request by lialan**
Previously when `numFrontPadElems` is not zero, `getCompressedMaskOp` produces wrong result if the mask generator op is a `vector.create_mask`. 

This patch resolves the issue by including `numFrontPadElems` into the mask generation.

### CODE REVIEW

**Review by dcaballe — APPROVED** [review]
LG, thanks! Please give some time for others to review.

File: mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir
Comment (banach-space) [inline mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:81]: [nit] From what I can tell, this is the first set of tests with `vector.create_mask` in this file? If yes, then I would skip "check" in func name (adds no info) and instead re-use a name of an exisiting test that this is a variation of. For example: `@check_unaligned_create_mask_dynamic_i2` --> `@vector_create_mask_maskedload_i2` or sth similar.

In fact, I'd rename `@vector_cst_maskedload_i2` as `@vector_constant_mask_maskedload_i2` and then you'd have consistent naming scheme that would allow easy identification of the tested corner cases.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:111]: I don't quite follow this comment. Is it meant to explain where the definition of `s0` is coming from? Or, more specifically:

> the compressed mask generated by shifting the original mask

where does shifting happen?

**Review by banach-space — COMMENTED** [review]
The updated logic makes sense, thanks! A few small suggestions.

File: mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir
Comment (lialan) [reply mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:80]: correct. Before it was `floordiv(m + (4 - 1), 4)`.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp
Comment (lialan) [reply mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:111]: The shifting happens when you pad `numFrontPadElems` to the mask so each individual mask value points to its correct element.

The shifting (to the right) is done by adding `numFrontPadElems` to the `ceildiv` in the affine expression. namely:
`s0 = (s0 + numFrontPadElems).ceilDiv(numSrcElemsPerDest);`

File: mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir
Comment (lialan) [reply mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir]: Thanks for the suggestion! I like this. updated

**Review by hanhanW — APPROVED** [review]
Thanks Alan, it looks good to me. I'll help land the PR once we get an approval from @banach-space

File: mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir
Comment (banach-space) [inline mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:77]: [nit] The input (vector and memrefs) are static. Only the indices are dynamic. It's worth clarifying.

**Review by banach-space — APPROVED** [review]
LGTM % a couple of nits

**Pull request merged**
